core/schema/modulesource.go

```diff
@@ -239,32 +260,84 @@ type buildkitClient interface {
 // - if not, try to isolate root of git repo from the ref
 // - if nothing worked, fallback as local ref, as before
 func parseRefString(ctx context.Context, bk buildkitClient, refString string) parsedRefString {
-	var parsed parsedRefString
-	parsed.modPath = refString
+	localParsed := parsedRefString{
+		modPath: refString,
+		kind:    core.ModuleSourceKindLocal,
+		scheme:  core.NoScheme,
+	}
 
-	// We do a stat in case the mod path github.com/username is a local directory
-	stat, err := bk.StatCallerHostPath(ctx, parsed.modPath, false)
+	// First, we stat ref in case the mod path github.com/username is a local directory
+	stat, err := bk.StatCallerHostPath(ctx, refString, false)
 	if err == nil && stat.IsDir() {
-		parsed.kind = core.ModuleSourceKindLocal
-		return parsed
+		return localParsed
 	}
 
-	parsed.modPath, parsed.modVersion, parsed.hasVersion = strings.Cut(refString, "@")
+	// Parse scheme and attempt to parse as git endpoint
+	gitParsed, err := parseGitEndpoint(refString)
+	if err == nil {
+		// Try to isolate the root of the git repo
+		repoRoot, err := vcs.RepoRootForImportPath(gitParsed.modPath, false)
+		if err == nil && repoRoot != nil && repoRoot.VCS != nil && repoRoot.VCS.Name == "Git" {
+			gitParsed.repoRoot = repoRoot
+			// the extra "/" trim is important as subpath traversal such as /../ are being cleaned by filePath.Clean
+			gitParsed.repoRootSubdir = strings.TrimPrefix(strings.TrimPrefix(gitParsed.modPath, repoRoot.Root), "/")
+			return gitParsed
+		}
+	}
 
-	// we try to isolate the root of the git repo
-	repoRoot, err := vcs.RepoRootForImportPath(parsed.modPath, false)
-	if err == nil && repoRoot != nil && repoRoot.VCS != nil && repoRoot.VCS.Name == "Git" {
-		parsed.kind = core.ModuleSourceKindGit
-		parsed.repoRoot = repoRoot
-		parsed.repoRootSubdir = strings.TrimPrefix(parsed.modPath, repoRoot.Root)
-		// the extra "/" is important as subpath traversal such as /../ are being cleaned by filePath.Clean
-		parsed.repoRootSubdir = strings.TrimPrefix(parsed.repoRootSubdir, "/")
-		return parsed
+	// Fallback to local reference
+	return localParsed
+}
+
+func parseGitEndpoint(refString string) (parsedRefString, error) {
+	scheme, schemelessRef := parseScheme(refString)
+
+	// we append "ssh" to ensure that NewEndPoint properly parses the host / path and user
+	// HTTP refs are valid SSH refs. Bonus: NewEndpoint library handles correctly implicit SSH refs
+	endpoint, err := transport.NewEndpoint("ssh://" + schemelessRef)
```

Comment:
Hmmm? This is a bit confusing. Why do we try and extract an `sshusername` if the scheme wasn't for `ssh://`?